Split up AppInit2 into multiple phases, daemonize after datadir lock errors

laanwj opened the pull request:
Split up AppInit2 into multiple steps: This allows doing some of the steps before daemonization and some later.

To avoid the issue described in https://github.com/bitcoin/bitcoin/issues/9009#issuecomment-255988169:
- Before daemonization, just probe the data directory lock and print an early error message if possible.
- After daemonization get the data directory lock again and hold on to it until exit. This creates a slight window for a race condition to happen, however this condition is harmless: it will at most make the process exit without printing a message to console as it will be detected after daemonization.

Solves #9009:

```
$ src/bitcoind -testnet -daemon
Bitcoin server starting
$ src/bitcoind -testnet -daemon
Error: Cannot obtain a lock on data directory /home/orion/.bitcoin/testnet3. Bitcoin Core is probably already running.
```

As for the AppInit2 split: This is something I've been wanting to do for a while. Note that, to reduce risk of regressions, this is just mechanic code movement and doesn't change ordering, besides the `daemon()` invocation.

Some smarter things could be done such as:
- Move basic context initialization from `AppInitBasicSetup()` to `SetupEnvironment()` where possible, especially things such as the windows DEP activation
- Unify `InitParameterInteraction()` and `AppInitParameterInteraction()` if/where possible. I'm not sure why this is done in two steps with just `AppInitBasicSetup()` in between right now.
- Do something smarter than `// Variables internal to initialization process only` (e.g. an initialization context structure). Some variables there may be completely avoided by moving the code to extract the arguments.
- Splitting up further steps, this will allow factoring out `ENABLE_WALLET` sections from `libbitcoin_server.a` (which contains init.cpp) and remove the circular dependencies (#7965)
- It is now possible to do unit testing on our initialization steps in isolation.

However I do not intend to do these things in this pull.

laanwj committed:
init: Split up AppInit2 into multiple phases

This allows doing some of the steps before e.g. daemonization and some
fater.

laanwj committed:
init: Try to aquire datadir lock before and after daemonization

Before daemonization, just probe the data directory lock and print an
early error message if possible.

After daemonization get the data directory lock again and hold on to it until exit
This creates a slight window for a race condition to happen, however this condition is harmless: it
will at most make us exit without printing a message to console.

    $ src/bitcoind -testnet -daemon
    Bitcoin server starting
    $ src/bitcoind -testnet -daemon
    Error: Cannot obtain a lock on data directory /home/orion/.bitcoin/testnet3. Bitcoin Core is probably already running.

jonasschnelli commented:
Code Review utACK feaea7001fdec6c0ec5ef6cc1b33cb214131ce25.
And yes, I think we should unify `InitParameterInteraction` and AppInitParameterInteraction() to avoid confusion (can be done in a later PR).

gmaxwell commented:
ACK.

sipa commented:
utACK feaea7001fdec6c0ec5ef6cc1b33cb214131ce25

maflcko reviewed (approved):
utACK feaea7001fdec6c0ec5ef6cc1b33cb214131ce25

MarcoFalke commented on src/init.cpp:
```diff
@@ -977,8 +988,7 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
 
-    ServiceFlags nLocalServices = NODE_NETWORK;
-    ServiceFlags nRelevantServices = NODE_NETWORK;
+    nLocalServices = NODE_NETWORK;
```
Any reason the default value was not assigned in the `namespace` (like `nRelevantServices`)?

laanwj replied:
No reason, will do that.

morcos commented:
@laanwj I don't think you're setting `fServer` correctly.  I think you need to move the `SoftSetBoolArg("-server", true)` to before `AppInitParameterInteraction`.

laanwj commented:
> @laanwj I don't think you're setting fServer correctly. I think you need to move the SoftSetBoolArg("-server", true) to before AppInitParameterInteraction.

There isn't any parameter interaction based on `-server`, so I don't think this makes a difference in practice. It'd be more logical to move it though, so I will.

morcos commented:
@laanwj `fServer` gets set in `AppInitParameterInteraction`.  I compiled this PR and ran `src/bitcoind -daemon` and was unable to connect to bitcoind using bitcoin-cli.

laanwj commented:
> @laanwj fServer gets set in AppInitParameterInteraction. I compiled this PR and ran src/bitcoind -daemon and was unable to connect to bitcoind using bitcoin-cli.

Oh you're right, sorry.
I expected the RPC tests to catch this but apparently they set `-server` explicitly.
```
qa/rpc-tests/test_framework/util.py:            args = [ os.getenv("BITCOIND", "bitcoind"), "-server", "-keypool=1", "-datadir="+datadir, "-discover=0" ]
qa/rpc-tests/test_framework/util.py:    args = [ binary, "-datadir="+datadir, "-server", "-keypool=1", "-discover=0", "-rest", "-mocktime="+str(get_mocktime()) ]
```
> EDIT: Perhaps the right solution is to remove fServer

It exists to make running a RPC server optional (disabled by default) with the GUI. I think we should keep that.

morcos commented:
Yes I just meant get rid of the variable `fServer` since its used only once and in `AppInitMain` just check `GetBoolArg("-server", false)` directly, but either way.

laanwj committed:
init: Get rid of fServer flag

There is no need to store this flag globally, the variable is only used
inside the initialization process.

Thanks to Alex Morcos for the idea.

laanwj commented:
> Yes I just meant get rid of the variable fServer since its used only once and in AppInitMain just check GetBoolArg("-server", false) directly, but either way.

Good idea. I pushed these changes:
- Initialize nRelevantServices in the namespace like nLocalServices (MarcoFalke's comment)
- Move `SetSoftBool("-server")` to before parameter interaction in `bitcoind.cpp`, in case someone adds parameter interaction based on this
- Get rid of global fServer flag

morcos commented:
ACK deec83f

MarcoFalke commented:
reACK deec83f. Good finding @morcos!
